lib/sycamore/sycamore/functions/chunker.py

```diff
@@ -1,41 +1,86 @@
+            Must be less than chunk_token_count.
+            
+    Raises:
+        ValueError: If chunk_overlap_token_count >= chunk_token_count or if either parameter is negative.
+        
     Example:
-         .. code-block:: python
-
-            chunker = TextOverlapChunker(chunk_token_count=1000, chunk_overlap_token_count=100)
-            chunks = chunker.chunk(data)
+        >>> chunker = TextOverlapChunker(chunk_token_count=5, chunk_overlap_token_count=2)
+        >>> tokens = list("ABCDEFGHIJK")
+        >>> chunks = chunker.chunk(tokens)
+        >>> for chunk in chunks: print(''.join(chunk))
+        ABCDE
+        DEFGH
+        GHIJK
     """
-
     def __init__(self, chunk_token_count: int = 1000, chunk_overlap_token_count: int = 100) -> None:
         super().__init__()
+        if chunk_token_count <= 0:
+            raise ValueError("Chunk token count must be positive")
+        if chunk_overlap_token_count < 0:
+            raise ValueError("Chunk overlap token count must be non-negative")
         if chunk_overlap_token_count >= chunk_token_count:
-            raise Exception("Token overlap count between chunks must be lesser than chunk token count")
+            raise ValueError("Token overlap count between chunks must be less than chunk token count")
+            
         self._chunk_token_count = chunk_token_count
         self._chunk_overlap_token_count = chunk_overlap_token_count
-
-    def chunk(self, tokens: list[Any]) -> list[Any]:
-        return [
-            tokens[a : a + self._chunk_token_count]
-            for a in range(0, len(tokens), self._chunk_token_count - self._chunk_overlap_token_count)
-        ]
+        
+    def chunk(self, tokens: List[T]) -> List[List[T]]:
+        """
+        Divide the input sequence into overlapping chunks.
+        
+        Args:
+            tokens: The input sequence to be chunked.
+            
+        Returns:
+            A list of chunks, where each chunk is a list of tokens.
+            
+        Note:
+            The last chunk may be smaller than chunk_token_count but will maintain
+            the specified overlap with the previous chunk if possible.
+        """
+        if not tokens:
+            return []
+            
+        chunks = []
+        stride = self._chunk_token_count - self._chunk_overlap_token_count
+        
+        for start in range(0, len(tokens), stride):
+            # Calculate end index for current chunk
+            end = min(start + self._chunk_token_count, len(tokens))
+            chunk = tokens[start:end]
+            
+            # Add chunk if it's the first chunk, maintains minimum size, or is the last piece
+            if (start == 0 or 
+                len(chunk) >= self._chunk_overlap_token_count or 
+                end == len(tokens)):
+                chunks.append(chunk)
+                
+            # If we've processed all tokens, break
+            if end == len(tokens):
+                break
```

Comment:
we can just exit the for loop because we're at the end of the for loop in this case, right?

Reply:
edit: I think this is the cause of the test failures - if the natural chunk boundary lines up exactly with the end of the sequence then you'll break prematurely, right?
overlap = 2, len = 5
"ABCDEFGH" -> "ABCDE", "DEFGH" instead of "ABCDE", "DEFGH", "GH"